src/ApiCompat/Microsoft.DotNet.ApiCompatibility/Rules/CannotChangeVisibility.cs

```diff
@@ -0,0 +1,94 @@
+﻿// Copyright (c) .NET Foundation and contributors. All rights reserved.
+// Licensed under the MIT license. See LICENSE file in the project root for full license information.
+
+
+using System.Collections.Generic;
+using System.Runtime;
+using Microsoft.CodeAnalysis;
+using Microsoft.DotNet.ApiCompatibility.Abstractions;
+using Microsoft.DotNet.ApiCompatibility.Extensions;
+
+namespace Microsoft.DotNet.ApiCompatibility.Rules
+{
+    /// <summary>
+    /// This class implements a rule to check that the visibility of symbols is not reduced.
+    /// In strict mode, it also checks that the visibility isn't expanded.
+    /// </summary>
+    public class CannotChangeVisibility : IRule
+    {
+        private readonly RuleSettings _settings;
+
+        public CannotChangeVisibility(RuleSettings settings, IRuleRegistrationContext context)
+        {
+            _settings = settings;
+            context.RegisterOnMemberSymbolAction(RunOnMemberSymbol);
+            context.RegisterOnTypeSymbolAction(RunOnTypeSymbol);
+        }
+
+        private void RunOnTypeSymbol(
+            ITypeSymbol? left,
+            ITypeSymbol? right,
+            MetadataInformation leftMetadata,
+            MetadataInformation rightMetadata,
+            IList<CompatDifference> differences)
+        {
+            if (left is null && right is null)
```

Comment:
If something goes from Public to Private, would you rather a diagnostic that says the member is missing, or that its visibility changed?

Reply:
If I'm not mistaken, with this change two diagnostics would be emitted: one for MemberMustExist and another for CaanotChangeVisibility. I think having both isn't what we want.

Reply:
This rule is meant to warn users when a member goes from more visible to less visible. For example, a method going from `public` to `protected`, or a type going from `public` to `internal`.

An example from the old APICompat
```
CannotMakeMoreVisible : Visibility of member 'System.Diagnostics.Tracing.EventListener.EventSourceIndex(System.Diagnostics.Tracing.EventSource)' is 'protected' in the reference but 'public' in the implementation.
```

However, I might have overreached in this PR to work on private symbols as well.

Reply:
There's one special case of this rule. Reducing the visibility of a protected member inside sealed type is permitted (since they're effectively private), but we can't really test for it, since we ignore private members. @ericstj suggested that ignoring "effectively private" members in the RuleRunner's traversal might be a better option.

Reply:
Right - it's more of a traversal problem, since you want to ignore *everything* about a protected member in a sealed type, not just the results of one rule check.

Reply:
If effectively sealed members would be excluded from the traversal, would that impact the existing logic that already treats such members in a special way? https://github.com/dotnet/sdk/search?q=effectively+sealed

Reply:
The abstract member case would not be affected, since it's trying to ignore members inside effectively sealed types. The CannotSealType doesn't care about members, so it would not be affected either.